spinalcordtoolbox/labels.py

```diff
@@ -224,52 +227,46 @@ def label_regions_from_reference(img: Image, ref: Image, centerline: bool = Fals
     :param centerline: boolean, if True the centerline is returned instead of the segmentation (default to false)
     :returns: segmentation image with vertebral levels labelized
     """
-
-    og_img_orientation = img.orientation
-    if og_img_orientation != "RPI":
-        img.change_orientation("RPI")
-
-    og_ref_orientation = ref.orientation
-    if og_ref_orientation != "RPI":
-        ref.change_orientation("RPI")
-
-    out = zeros_like(img)
-
-    # Extract centerline from segmentation
-    _, arr_ctl, _, _ = get_centerline(img)
-    centerline_arr = arr_ctl.T
-
+    # The centerline of the segmentation, with one point for each S-I slice.
+    centerline_pix = get_centerline(img, space='pix')[1].T.astype(int)  # shape (N, 3), int
+    centerline_phys = img.transfo_pix2phys(centerline_pix, mode='absolute')  # shape (N, 3), float
+
+    # Prepare to compute, for several points, the centerline index which is closest in physical space.
+    centerline_tree = KDTree(centerline_phys)
+
+    # The label coordinates.
+    labels_pix = np.argwhere(ref.data)  # shape (L, 3), int
+    labels_phys = ref.transfo_pix2phys(labels_pix, mode='absolute')  # shape (L, 3), float
+    # The corresponding label values.
+    labels_values = ref.data[tuple(labels_pix.T)]  # shape (L,), arbitrary dtype
+    # The closest centerline index for each label voxel.
+    _, labels_indices = centerline_tree.query(labels_phys)  # shape (L,), int
+
+    # Sort everything by increasing centerline index
+    sorter = np.argsort(labels_indices)
+    labels_pix = labels_pix[sorter]
+    labels_phys = labels_phys[sorter]
```

Comment:
This two quantities don't actually get used elsewhere in the function, right? (I guess it's useful to still sort them to aid with debugging, since the internal state at this point will stay consistent between these quantities.)

Reply:
Yes, I had this same debate internally. Having a clear, consistent state for debugging is also why I did `del sorter` at the end of this paragraph: to really signal that we're done using this permutation.